pandas/core/indexes/base.py

```diff
@@ -3111,27 +3111,64 @@ def join(self, other, how='left', level=None, return_indexers=False,
 
     def _join_multi(self, other, how, return_indexers=True):
         from .multi import MultiIndex
-        self_is_mi = isinstance(self, MultiIndex)
-        other_is_mi = isinstance(other, MultiIndex)
+        from pandas.core.reshape.merge import _complete_outer_join
 
         # figure out join names
         self_names = [n for n in self.names if n is not None]
         other_names = [n for n in other.names if n is not None]
-        overlap = list(set(self_names) & set(other_names))
+        overlap = set(self_names) & set(other_names)
 
-        # need at least 1 in common, but not more than 1
+        # need at least 1 in common
         if not len(overlap):
-            raise ValueError("cannot join with no level specified and no "
-                             "overlapping names")
-        if len(overlap) > 1:
-            raise NotImplementedError("merging with more than one level "
-                                      "overlap on a multi-index is not "
-                                      "implemented")
-        jl = overlap[0]
-
-        # make the indices into mi's that match
+            raise ValueError("cannot join with no overlapping index names")
+
+        # Drop the non matching levels
+        ldrop_levels = list(set(self_names) - overlap)
+        rdrop_levels = list(set(other_names) - overlap)
+
+        self_is_mi = isinstance(self, MultiIndex)
+        other_is_mi = isinstance(other, MultiIndex)
+
+        if self_is_mi and other_is_mi:
+            self_tmp = self.droplevel(ldrop_levels)
+            other_tmp = other.droplevel(rdrop_levels)
+
+            if not (self_tmp.is_unique and other_tmp.is_unique):
+                raise ValueError("Join on level between two MultiIndex objects"
+                                 "is ambiguous")
+
+            # Append to the returned Index the non-overlapping levels
+            dropped_lvls = ldrop_levels + rdrop_levels
+
+            # tmp_index is equivalent of index when how='inner'
+            tmp_index, lidx, ridx = self_tmp.join(other_tmp, how=how,
```

Comment:
don't use ``tmp_index`` call this something else (and self_tmp)

Reply:
How about:
self_tmp --> left
other_tmp --> right
tmp_index --> inner_join_index